### PR TITLE
nixos system-activation: support activation scripts run in a user context

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

#47842 for 18.09

Cc: @samueldr @vcunat 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by samueldr — APPROVED** [review]
Same as the other PR; though verified and it works as expected with a custom activation script snippet.

File: nixos/modules/services/x11/desktop-managers/plasma5.nix [hunk 1]
```diff
@@ -225,11 +225,8 @@ in
       security.pam.services.sddm.enableKwallet = true;
       security.pam.services.slim.enableKwallet = true;
 
-      # Update the start menu for each user that has `isNormalUser` set.
-      system.activationScripts.plasmaSetup = stringAfter [ "users" "groups" ]
-        (concatStringsSep "\n"
-          (mapAttrsToList (name: value: "${pkgs.su}/bin/su ${name} -c ${pkgs.libsForQt5.kservice}/bin/kbuildsycoca5")
-            (filterAttrs (n: v: v.isNormalUser) config.users.users)));
+      # Update the start menu for each user that is currently logged in
+      system.userActivationScripts.plasmaSetup = "${pkgs.libsForQt5.kservice}/bin/kbuildsycoca5";
```
Comment (grahamc) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix]: Why use a user-activation script instead of just creating a user systemd unit?

Comment (vcunat) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: You missed that `system.userActivationScripts` is implemented via a systemd unit?

**Comment by grahamc** [comment]
No, I wonder why have an “activation script” not part of the “activate” script, instead of just having these modules create their own use case specific systemd unit.

**Comment by vcunat** [comment]
The system activation script is ran once and with system privileges.  This is to be ran for each (active) user with their privileges.

**Comment by grahamc** [comment]
Right. Based on the definition it seems it will run on every login. I’m just not sure why we need the concept of a user activation script when a one shot user unit works just the same. 

This is in contrast to system activation scripts which aren’t typically implementable as systemd units.

**Comment by peterhoeg** [comment]
> Right. Based on the definition it seems it will run on every login. 

That's not entirely true. The system level activation scripts do run on every boot, but the user context scripts are only triggered for users logged in (we get a list of active sessions from logind).

> I’m just not sure why we need the concept of a user activation script when a one shot user unit works just the same. 

That's how it's implemented - as a oneshot unit. We still need a way to trigger it though so that is what the system level activation script handles.

> This is in contrast to system activation scripts which aren’t typically implementable as systemd units.

Well, it's actually moving that way - @jameysharp has done some nice work in https://github.com/NixOS/nixpkgs/pull/47563.

**Comment by grahamc** [comment]
Nice. I understand now. Thank you!

**Comment by jameysharp** [comment]
Ooh, I like it. Although I think I'd like it even better if `nixos-activation.service` were instead `nixos-activation.target`. Then instead of `system.userActivationScripts`, you'd have `systemd.user.services` with `wantedBy = [ "nixos-activation.target" ];`.

**Comment by peterhoeg** [comment]
@jameysharp  - I'm not ignoring you. I know how disheartening it can be when putting in a lot of thought into something, writing it up and then to be met with silence, but time's been a little tight on this side. I want us to work on getting something solid out of this and will get back to you shortly.

**Pull request merged**